### PR TITLE
[RFC] API for request the bootloader mode by the application 

### PR DESCRIPTION
**Pull request by nvlsianpu**
This API is added in order to make possible entering given mode by the bootloader.

Particular user case is entering the DFU recovery mode by MCUboot if application has requested that.

This RFC is created as result from discussion on [the retention register API proposla](https://github.com/zephyrproject-rtos/zephyr/pull/50105).
Idea is to unblock possibility for the bootloader mode request feature.
see [architecture WG comment](https://github.com/zephyrproject-rtos/zephyr/pull/50105#issuecomment-1259701243)

Implementation of this might be reworked once common API for (used here) retention register will be established.
Similar remark is valid for nRF specific `int nrf_sys_get_reboot_type(void)` - I might be generalized as well in the feature, but it is no the call to do that.

Previous attempts to introduce similar API can be fund there:
#46084
#44512

### CODE REVIEW

File: subsys/dfu/Kconfig [hunk 1]
```diff
@@ -96,3 +97,30 @@ config UPDATEABLE_IMAGE_NUMBER
 endif
 
 endif # IMG_MANAGER
+
+menuconfig BOOT_MODE_API
+	bool "API for request bootloader mode"
```
Comment (nordicjm) [inline subsys/dfu/Kconfig:102]: *requesting

File: subsys/dfu/Kconfig [hunk 2]
```diff
@@ -96,3 +97,30 @@ config UPDATEABLE_IMAGE_NUMBER
...
+	help
+	  Application can request specific bootloader mode. The bootloader might
+	  enter this mode once system reboot usally.
+
+if BOOT_MODE_API
+choice
+	prompt "bootloader mode API implementation"
```
Comment (nordicjm) [inline subsys/dfu/Kconfig:109]: *Bootloader

File: subsys/dfu/Kconfig [hunk 3]
```diff
@@ -96,3 +97,30 @@ config UPDATEABLE_IMAGE_NUMBER
...
+	default BOOT_MODE_API_MCUBOOT
+	help
+	  Choice support for managing DFU image.
```
Comment (nordicjm) [inline subsys/dfu/Kconfig:112]: Comment does not seem applicable to this option

File: soc/arm/nordic_nrf/nrf52/soc.c [hunk 4]
```diff
@@ -69,6 +72,20 @@ uint8_t soc_retention_reg_mask_get(uint8_t mask, uint8_t pos)
...
+	soc_retention_reg_mask_set(mode, SOC_MCUBOOT_MODE_Msk,
+				   SOC_MCUBOOT_MODE_Pos);
+
+	return 0;
+}
+
+uint8_t soc_mcuboot_mode_get(void)
+{
+	return soc_retention_reg_mask_get(SOC_MCUBOOT_MODE_Msk,
+				      SOC_MCUBOOT_MODE_Pos);
```
Comment (nordicjm) [inline soc/arm/nordic_nrf/nrf52/soc.c:86]: Alignment

File: soc/arm/nordic_nrf/nrf52/soc.c [hunk 5]
```diff
@@ -47,11 +47,34 @@ extern void z_arm_nmi_init(void);
 #define LOG_LEVEL CONFIG_SOC_LOG_LEVEL
 LOG_MODULE_REGISTER(soc);
 
+#define SOC_MCUBOOT_RESET_TYPE_Pos (0UL)
```
Comment (nordicjm) [inline soc/arm/nordic_nrf/nrf52/soc.c:50]: Needs documentation to explain what bits are used/reserved (for what features/Kconfigs) and which bits the user can use for their own purposes

File: soc/arm/nordic_nrf/common/soc_nrf_common.h [hunk 6]
```diff
@@ -236,6 +236,21 @@
 		DT_NODE_PATH(node_id) " has pinctrl states defined"	\
 			" although PINCTRL is not enabled")
 
+/**
+ * @brief Obtain SW defined reboot type
+ *
+ * Fetch reboot type which caused system startup.
+ * This call extract reboot type vale which was used in previous call to the
+ * sys_rebot() call if the platform allow to store the value.
```
Comment (nordicjm) [inline soc/arm/nordic_nrf/common/soc_nrf_common.h:244]:
```suggestion
 * This call extracts the reboot type value which was used in a previous call to
 * sys_rebot() if the platform allows storing of this.
```

File: soc/arm/nordic_nrf/common/soc_nrf_common.h [hunk 7]
```diff
@@ -236,6 +236,21 @@
...
+ *
+ * @note This function is intended to be replaced in the feature when better
+ *	 common API will come. This nRF specific API is needed as temporaly
```
Comment (nordicjm) [inline soc/arm/nordic_nrf/common/soc_nrf_common.h:247]:
```suggestion
 *	 common API will come. This nRF specific API is needed as temporarily
```

File: soc/arm/nordic_nrf/nrf52/soc.c [hunk 8]
```diff
@@ -50,6 +50,9 @@ LOG_MODULE_REGISTER(soc);
 #define SOC_MCUBOOT_RESET_TYPE_Pos (0UL)
 #define SOC_MCUBOOT_RESET_TYPE_Msk (0x1FUL << SOC_MCUBOOT_RESET_TYPE_Pos)
 
+#define SOC_MCUBOOT_MODE_Pos (5UL)
+#define SOC_MCUBOOT_MODE_Msk (0x7UL << SOC_MCUBOOT_MODE_Pos)
```
Comment (nordicjm) [inline soc/arm/nordic_nrf/nrf52/soc.c:54]: Why does this need 3 bits?

**Comment by carlescufi** [comment]
@de-nordic could you please review this?

**Review by mbolivar-nordic — COMMENTED** [review]
I understand the use case but I agree with the other comments about this feeling a bit under-documented. When I read our product specifications, I don't really see any information that GPREGRET is reserved for platform specific use, which is what we seem to be doing here. The whole thing feels kind of magical in a way that isn't quite nice to me.

That said, I don't know the goals here well enough to say that this should not go in, so I'm neither +1 nor -1.

**Comment by nvlsianpu** [comment]
@nordicjm @de-nordic You can takeover this if you want.

**Pull request closed without merging**